### PR TITLE
WAF extraction

### PR DESCRIPTION
**Pull request by rshewitt**
# Pull Request

Related to [#4558](https://github.com/GSA/data.gov/issues/4558)

## About

function for traversing a WAF and collecting all the .xml files in the tree.

## PR TASKS

- [x] The actual code changes.
- [x] Tests written and passed.
- [ ] Any changes to docs?
- [ ] Bumped version number in [pyproject.toml](https://github.com/GSA/datagov-harvesting-logic/blob/8078c5b9f015ca7fb8a142e4e4b2e22ad2fd49b1/pyproject.toml#L3) (also checked on [PyPi](https://pypi.org/project/ckanext-geodatagov/#history)).

### CODE REVIEW

**Comment by rshewitt** [comment]
WAF tested against
```
localhost

└───waf
│   │   USGSHydroCached.xml
│   │   USGSImageryOnly.xml
│   │   USGSImageryTopo.xml
│   │   USGSShadedReliefOnly.xml
│   │
│   |───browse
│   |    |   USGSHydroCached.xml
|   |    |
│   |    └───more
│   |            USGSShadedReliefOnly.xml 
|   |    
│   └───other
│          USGSImageryOnly.xml
│   
└───dcatus
```

File: harvester/extract/waf.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import requests
+from bs4 import BeautifulSoup
+import os
+
+
+def traverse_waf(url, files=[], file_ext=".xml", folder="/", filters=[]):
+    parent = os.path.dirname(url.rstrip("/"))
+
+    res = requests.get(url)
+    if res.status_code == 200:
```
Comment (btylerburton) [inline harvester/extract/waf.py]: can you add some pseudo code, or a TODO, for error handling?

**Review by btylerburton — COMMENTED** [review]
looks great for a first pass. my only concern is that it creates merge conflicts with @robert-bryson's work on this ticket: https://github.com/GSA/data.gov/issues/4544

so long as you coordinate your changes, though we should be good...

File: harvester/extract/waf.py (hunk 1 above)
Comment (rshewitt) [reply harvester/extract/waf.py]: [added](https://github.com/GSA/datagov-harvesting-logic/pull/24/commits/aee51a627bc2ce42387c8131996c14b0ccdb82e9)

**Review by jbrown-xentity — DISMISSED** [review]
Looks good!

File: harvester/extract/waf.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+def traverse_waf(url, files=[], file_ext=".xml", folder="/", filters=[]):
+    # TODO: add exception handling
+    parent = os.path.dirname(url.rstrip("/"))
+
+    res = requests.get(url)
+    if res.status_code == 200:
+        soup = BeautifulSoup(res.content, "html.parser")
+        anchors = soup.find_all("a", href=True)
+
+        folders = []
```
Comment (btylerburton) [inline harvester/extract/waf.py]: After so many years in JS land, Python function scoping always gives me pause.

**Review by btylerburton — DISMISSED** [review]
Looks great!

**Review by robert-bryson — APPROVED** [review]
Great work!

**Pull request merged**